plugins/reindex/src/main/java/org/elasticsearch/plugin/reindex/AbstractAsyncBulkByScrollAction.java

```diff
@@ -0,0 +1,352 @@
+
+    public void start() {
+        initialSearch();
+    }
+
+    /**
+     * Count of documents updated.
+     */
+    public long updated() {
+        return updated.get();
+    }
+
+    /**
+     * Count of documents created.
+     */
+    public long created() {
+        return created.get();
+    }
+
+    /**
+     * Count of successful delete operations.
+     */
+    public long deleted() {
+        return deleted.get();
+    }
+
+    /**
+     * The number of scan responses this request has processed.
+     */
+    public int batches() {
+        return batches.get();
+    }
+
+    public long versionConflicts() {
+        return versionConflicts.get();
+    }
+
+    public long successfullyProcessed() {
+        return updated.get() + created.get() + deleted.get();
+    }
+
+    public List<Failure> failures() {
+        return unmodifiableList(failures);
+    }
+
+    void initialSearch() {
+        try {
+            startTime.set(System.nanoTime());
+            if (logger.isDebugEnabled()) {
+                logger.debug("executing initial scroll against {}{}",
+                        firstSearchRequest.indices() == null ? "all indices" : firstSearchRequest.indices(),
+                        firstSearchRequest.types() == null || firstSearchRequest.types().length == 0 ? "" : firstSearchRequest.types());
+            }
+            client.search(firstSearchRequest, new ActionListener<SearchResponse>() {
+                @Override
+                public void onResponse(SearchResponse response) {
+                    logger.debug("[{}] documents match query", response.getHits().getTotalHits());
+                    total.set(response.getHits().getTotalHits());
+                    onScrollResponse(response);
+                }
+
+                @Override
+                public void onFailure(Throwable e) {
+                    logger.error("failed while executing the initial scroll request", e);
+                    listener.onFailure(e);
+                }
+            });
+        } catch (Throwable t) {
+            finishHim(t);
+        }
+    }
+
+    void onScrollResponse(SearchResponse searchResponse) {
+        threadPool.executor(ThreadPool.Names.GENERIC).execute(() -> {
+            try {
+                scroll.set(searchResponse.getScrollId());
+                SearchHit[] docs = searchResponse.getHits().getHits();
+                logger.debug("scroll returned [{}] documents with a scroll id of [{}]", docs.length, searchResponse.getScrollId());
+                if (docs.length == 0) {
+                    startNormalTermination();
+                    return;
+                }
+                batches.incrementAndGet();
+                List<SearchHit> docsIterable = Arrays.asList(docs);
+                if (mainRequest.size() != -1) {
+                    // Truncate the docs if we have more than the request size
+                    long remaining = max(0, mainRequest.size() - successfullyProcessed());
```

Comment:
Maybe a matter of personal taste (and I have not checked what's typical in the code base) but I think static imports on classes like `Collections` or `Math` make the code a bit harder to read for me (I've already wondered whether you have implemented `#max()` in this class).